### DISCUSSION TITLE
Porting Term into Comonicon printing system

### DISCUSSION THREAD

**Discussion by camilogarciabotero** [FedeClaudi/Term.jl#193] [Ideas]
Hi @FedeClaudi

I was unsure whether to open the discussion here or in the `Comonicon.jl` repo, but anyways...

Since https://github.com/JuliaLang/julia/pull/44527 landed, it is certainly better to use Julia as a scripting language, then it naturally comes that command line applications (CLI) may be a thing in the coming versions. There is a lot of stuff related to CLI, but one thing that is pretty useful is the design of how the CLI will take the flags, arguments, and options. Here comes the [`Comonicon.jl`](https://comonicon.org/stable/) package (kudos to @Roger-luo!) it has many features and works lovely! Here is a simple example of how a function becomes a `main` command: 

```julia
using Comonicon

@main function mycmd(arg; option="Sam", flag::Bool=false)
    @show arg
    @show option
    @show flag
end
```

Will give this:

![](https://comonicon.org/stable/assets/images/myscript.png)

It is nice, and I was fine with it. However, a time ago I came across [Typer](https://typer.tiangolo.com) which is

>[...] a library for building CLI applications that users will love using and developers will love creating. Based on Python 3.6+ type hints.

Basically applying `rich` to Python to create gorgeous CLIs. here is an example of it:

```python
import typer


def main(name: str):
    print(f"Hello {name}")


if __name__ == "__main__":
    typer.run(main)
``` 

![Screenshot 2023-01-13 at 11 14 45 PM](https://user-images.githubusercontent.com/25586362/212450172-62b313a9-86c5-4af3-bde4-82b1765ccb5f.png)

I believe this is could be achieved using Term with that panel feature and will be awesome!

**Comment by FedeClaudi** [comment]
That's very interesting, and something Term could certainly do, Term's meant to be a counterpart to rich in Python for Julia. 
I'd be very happy to see this happen I could even help with the work on it. Would either you or @Roger-luo be interested in working on this?

**Comment by Roger-luo** [comment]
I probably won't have time to work on this, also I don't think we should have Comonicon depending on the Term given how heavy it is

tested on v1.8.5

```julia
julia> @time using Comonicon
  0.097803 seconds (309.89 k allocations: 21.233 MiB)

julia> @time using Term
  0.642835 seconds (1.78 M allocations: 122.588 MiB, 10.77% gc time, 0.24% compilation time)
```

however, one can write an extra code generator like other [code generators](https://github.com/comonicon/Comonicon.jl/tree/main/src/codegen), or we can provide a custom printer API for the Julia code generator and implement this as a separate package.

**Reply by Roger-luo** [reply]
I like the idea though it's just we don't have much latency budget in Julia to have richer built-ins. 

> Since https://github.com/JuliaLang/julia/pull/44527 landed, it is certainly better to use Julia as a scripting language, then it naturally comes that command line applications (CLI) may be a thing in the coming versions.

I think the next step is to wait for loading time optimizations, e.g lazy sub-packages, and static compilation before that happens writing CLI applications in Julia has to be very careful to avoid unnecessary latencies.

**Reply by FedeClaudi** [reply]
That's fair. 

Just a note, not sure what version of Term.jl you've used for the test above, but I've been working on improving pre-compilation so I'm hoping to bring down latency for Term.

**Reply by camilogarciabotero** [reply]
Adding some benches on `v.1.9.0-beta2` and `Term v.1.10`

```julia
julia> @time using Term
  1.517197 seconds (1.72 M allocations: 107.844 MiB, 4.51% gc time)
```

Second:

```julia
julia> @time using Term
  0.000519 seconds (184 allocations: 14.078 KiB)
```

I know that the second execution is not actually what happens when a CLI is used, but it is now quite cheap, isn't it? Could `Term v.2.0` be like the second benchmark? I switched to the master branch of `Term`, and tried in a fresh Julia session, and now it looks like the loading latency is quite good:

```julia
julia> @time using Term
  0.000351 seconds (184 allocations: 14.078 KiB)
```

**Comment by camilogarciabotero** [comment]
> I'd be very happy to see this happen I could even help with the work on it. 

Great, I was playing around with [`src/ast/print.jl`](https://github.com/comonicon/Comonicon.jl/blob/main/src/ast/printing.jl) API of `Comonicon` (which I thought could be a nice starting point) and have some troubles with `Term.Panel` (as I'm unfamiliar with `Comonicon` to use them well), yet I ended up using `hLine` and got some differences in the CLI:

![Screenshot 2023-01-15 at 7 25 01 PM](https://user-images.githubusercontent.com/25586362/212575664-879f2748-8668-49d4-a38c-e641c98c319a.png)

This was achieved by adding `hline` to what is defined as `section` in the printing system.

> I don't think we should have Comonicon depending on the Term given how heavy it is

Regarding the dependency issue of having `Term`, and adding to some of what you later suggested, what if we only use part of it, like `using Term: Panel` or `using Term:hLine`? Is this helpful in reducing startup latency? I haven't noticed heavy differences while making the aforementioned changes in my local machine. 

The other ideas you mentioned sound great, yet I'm not that skillful to write a `code generator`.

> however, one can write an extra code generator like other [code generators](https://github.com/comonicon/Comonicon.jl/tree/main/src/codegen), or we can provide a custom printer API for the Julia code generator and implement this as a separate package
